File: go/common/answer.go
package common

import "fmt"

func formatPrefix(year, day int, part *int) string {
	partStr := ""
	if part != nil {
		partStr = fmt.Sprintf(" [Part %d]", *part)
	}
	return fmt.Sprintf("[AOC %d] [Day %02d]%s", year, day, partStr)
}

func Answer[T any](year, day int, part *int, result T) {
	fmt.Printf("%s %v\n", formatPrefix(year, day, part), result)
}

func AnswerPart[T any](year, day, part int, result T) {
	Answer(year, day, &part, result)
}

func Verify[T comparable](year, day int, part *int, expected, actual T) {
	var status string
	if expected == actual {
		status = "PASS"
	} else {
		status = fmt.Sprintf("FAIL (expected: %v, actual: %v)", expected, actual)
	}
	fmt.Printf("%s %s\n", formatPrefix(year, day, part), status)
}

func VerifyPart[T comparable](year, day, part int, expected, actual T) {
	Verify(year, day, &part, expected, actual)
}

File: rust/src/common/answer.rs
use std::fmt::Display;

fn format_prefix(year: u32, day: u32, part: Option<u32>) -> String {
    match part {
        Some(p) => format!("[AOC {}] [Day {:02}] [Part {}]", year, day, p),
        None => format!("[AOC {}] [Day {:02}]", year, day),
    }
}

pub fn answer<T: Display>(year: u32, day: u32, part: Option<u32>, result: T) {
    println!("{} {}", format_prefix(year, day, part), result);
}

pub fn answer_part<T: Display>(year: u32, day: u32, part: u32, result: T) {
    answer(year, day, Some(part), result);
}

pub fn verify<T: Display + PartialEq>(
    year: u32,
    day: u32,
    part: Option<u32>,
    expected: T,
    actual: T,
) {
    let status = if expected == actual {
        "PASS".to_string()
    } else {
        format!("FAIL (expected: {}, actual: {})", expected, actual)
    };
    println!("{} {}", format_prefix(year, day, part), status);
}

pub fn verify_part<T: Display + PartialEq>(year: u32, day: u32, part: u32, expected: T, actual: T) {
    verify(year, day, Some(part), expected, actual);
}
